storage: handle replica in purgatory but not in queue

tbg opened the pull request:
I don't know if I'm missing something larger here. This code is all a
mess.

Fixes #36659.

Release note: None

tbg commented:
I ran `make stress STRESSFLAGS='-p 4' PKG=./pkg/ccl/partitionccl/ TESTS=TestRepartitioning` for 15 minutes without problems (that test also ran into this problem in the nightlies: https://github.com/cockroachdb/cockroach/issues/35487).

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36662#-:-Lc6JW1P--Dgx4Qju1eu:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell, @petermattis, and @tbg)

___
*[pkg/storage/queue.go, line 928 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36662#-Lc6JSMG2Nqg1htioELU:-Lc6JSMG2Nqg1htioELV:b2mw8v8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4cc06f7f2fa1edd4cae23740a6421589a16c050/pkg/storage/queue.go#L928)):*
> ```Go
> 					for rangeID := range bq.mu.purgatory {
> 						item, ok := bq.mu.replicas[rangeID]
> 						if ok {
> ```

Nit: this could be `item, ok := ...; if ok { ... }`


<!-- Sent from Reviewable.io -->

tbg committed:
storage: handle replica in purgatory but not in queue

I don't know if I'm missing something larger here. This code is all a
mess.

Fixes #36659.

Release note: None

tbg reviewed (commented):
TFTR!

bors r=petermattis

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36662#-:-Lc6JxMe6s4jeNfE92OD:b-spmsx3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bdarnell)



<!-- Sent from Reviewable.io -->
